### PR TITLE
fix: reworked full complement conditions + added _validate_recovery_period()

### PR DESCRIPTION
**Pull request by Inglezos**
## Related issues
This is related to #514 and #500.

## What was changed
Reworked full complement conditions by adding _validate_recovery_period() method in JHUDataComplementHandler class, in order to determine if the raw recovered data (for a specific country) is valid or not, in order to apply full complement. Invalid partial recovery periods (elapsed intervals in days) are considered if they are outside the range [7, 90] days and this behavior is exhibited by more than 50% of these elapsed intervals.

Now the countries with full complement are only 24:
['American Samoa', 'Belgium', 'Germany', 'Dominica', 'Fiji', 'France', 'United Kingdom', 'Grand Princess', 'Guam', 'Saint Kitts and Nevis', 'Laos', 'Marshall Islands', 'Northern Mariana Islands', 'MS Zaandam', 'Netherlands', 'Puerto Rico', 'Solomon Islands', 'Sweden', 'Timor-Leste', 'Holy See', 'Saint Vincent and the Grenadines', 'Virgin Islands, U.S.', 'Vanuatu', 'Samoa'].

To sum up:
1. China and Singapore are now not fully complemented.
2. UK, Germany, Netherlands and Sweden keep their full complement.
3. While France and Belgium now will be fully complemented, which is considered as more correct in my opinion, because for example the estimated Rt for France is very large (20-30) while now it lowers to normal values and both have unexpectedly low raw recovered data compared to total cases.

### CODE REVIEW

File: covsirphy/cleaning/jhu_complement.py [hunk 1]
```diff
@@ -145,6 +145,9 @@ def _pre_processing(self, subset_df):
                 Index: Date (pandas.TimeStamp)
                 Columns: Confirmed, Fatal, Recovered
         """
+        sel_valid_R = subset_df[self.C] - subset_df[self.F] < subset_df[self.R]
```
Comment (lisphilar) [inline covsirphy/cleaning/jhu_complement.py:148]: Just to confirm, do you mean "invalid" by `sel_valid_R`?

File: covsirphy/cleaning/jhu_complement.py [hunk 2]
```diff
@@ -162,6 +165,7 @@ def _post_processing(self, df):
                 Columns: Date (pandas.TimeStamp), Confirmed, Infected, Fatal, Recovered (int)
         """
         df = df.astype(np.int64)
+        df = df.loc[df[self.C] > self.max_ignored]
```
Comment (lisphilar) [inline covsirphy/cleaning/jhu_complement.py:168]: I think `df.loc[df[self.C] <= self.max_ignored]` should be added to the returned dataframe by `pd.concat().sort_index()` after line170.

File: covsirphy/cleaning/jhu_complement.py [hunk 3]
```diff
@@ -197,10 +201,53 @@ def _monotonic(self, df, variable):
             series.fillna(method="ffill", inplace=True)
             # Reduce values to the previous date
             df.loc[:date, variable] = series * raw_last / series.iloc[-1]
-            df[variable] = df[variable].fillna(0).astype(np.int64)
+            df[variable] = df[variable].fillna(0)
+        df[variable] = df[variable].apply(np.ceil).astype(np.int64)
         self.complement_dict[f"Monotonic_{variable.lower()}"] = True
         return df
 
+    def _validate_recovery_period(self, df, upper_limit_days=90, lower_limit_days=7,
```
[the remark this answers is not in the document]
Comment (lisphilar) [reply covsirphy/cleaning/jhu_complement.py:209]: Please move the arugments, including `upper_limit_days`, to `__init__` so that users can control them.

File: covsirphy/cleaning/jhu_complement.py [hunk 4]
```diff
@@ -224,10 +271,9 @@ def _recovered_full(self, df):
             sel_2 = df[self.C].diff().diff().rolling(14).mean() > 0
             cf_diff = (df[self.C] - df[self.F]).rolling(14).mean()
             sel_3 = df[self.R] < 0.01 * cf_diff
-            sel_4 = df[self.R] > 0.99 * cf_diff
-            s_df_1 = df.loc[sel_C1 & sel_2 & sel_4]
-            s_df_2 = df.loc[sel_R1 & sel_2 & sel_3]
-            if s_df_1.empty and s_df_2.empty:
+            s_df = df.loc[sel_C1 & sel_R1 & sel_2 & sel_3]
+            recovery_period_is_valid = self._validate_recovery_period(df)
+            if s_df.empty and recovery_period_is_valid:
```
Comment (lisphilar) [inline covsirphy/cleaning/jhu_complement.py:276]: Please replace line 375 and 276 with `if s_df.empty and self._validate_recovery_period(df):` to avoid calculation of `self._validate_recovery_period(df)` when `s_df.empty`.

File: covsirphy/cleaning/jhu_complement.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lisphilar) [reply covsirphy/cleaning/jhu_complement.py:209]: Please also update `JHUData.subset_complement()` and `JHUData.show_complement().`
To reduce the number of arguments of them, please replace `interval=2, max_ignored=100, max_ending_unupdated=14` with `**kwargs`.

In docsting, please add "kwargs: keyword arguments of JHUDataComplementHandler(), controll factors of complement".

File: covsirphy/cleaning/jhu_complement.py [hunk 5]
```diff
@@ -197,10 +201,53 @@ def _monotonic(self, df, variable):
...
+        check_df = check_df.unstack().reset_index()
+        check_df.columns = ["Variable", "Date", "Number"]
+        check_df["Days"] = (check_df[self.DATE] - check_df[self.DATE].min()).dt.days
+        # Calculate partial recovery periods
+        check_df = check_df.pivot_table(values="Days", index="Number", columns="Variable")
+        check_df = check_df.interpolate(limit_area="inside").dropna().astype(np.int64)
+        check_df["Elapsed"] = check_df[self.R] - check_df["diff"]
+        check_df = check_df.loc[check_df["Elapsed"] > 0]
+        # Check partial recovery periods
+        if sum(check_df["Elapsed"] > upper_limit_days)/len(check_df) > upper_percentage or \
```
Comment (lisphilar) [inline covsirphy/cleaning/jhu_complement.py:246]: Flake8 test failed.
Please replace
```Python
if sum(check_df["Elapsed"] > upper_limit_days)/len(check_df) > upper_percentage or \
           sum(check_df["Elapsed"] < lower_limit_days)/len(check_df) > lower_percentage:
            recovery_period_is_valid = False
        return recovery_period_is_valid
```
with 
```Python
per_up = (check_df["Elapsed"] > self.upper_limit_days).sum() / len(check_df)
per_lw = (check_df["Elapsed"] < self.lower_limit_days).sum() / len(check_df)
return per_up < self.upper_percentage and per_lw < self.lower_percentage
```
This change enable us to remove line 231 `recovery_period_is_valid = True`.

File: covsirphy/cleaning/jhu_complement.py [hunk 6]
```diff
@@ -197,10 +201,53 @@ def _monotonic(self, df, variable):
...
+
+        Returns:
+            bool: true if recovery period is within valid range or false otherwise
+
+        Note: Passed argument df corresponds to specific country's df
+              upper_limit_days has default value of 3 months (90 days)
+              lower_limit_days has default value of 1 week (7 days)
+        """
+        recovery_period_is_valid = True
+        check_df = df.copy()
```
Comment (lisphilar) [inline covsirphy/cleaning/jhu_complement.py:232]: I think it is unnecessary to copy it using memory, because we do not use the input dataframe later in this method.

**Review by lisphilar — CHANGES_REQUESTED** [review]
I added comments regarding minor issues. Kindly update codes.

File: covsirphy/cleaning/jhu_complement.py (hunk 1 above)
Comment (Inglezos) [reply covsirphy/cleaning/jhu_complement.py:148]: Yes invalid

File: covsirphy/cleaning/jhu_complement.py (hunk 2 above)
Comment (Inglezos) [reply covsirphy/cleaning/jhu_complement.py:168]: With 168 line I meant to completely ignore less than max_ignored cases. Essentially to concat them at the end of post processing means that in the first place this change shall not be done at all and thus remove the 168 line, right?

File: covsirphy/cleaning/jhu_complement.py (hunk 5 above)
Comment (Inglezos) [reply covsirphy/cleaning/jhu_complement.py:246]: We may spare the True value line but using this inverse logic is not very representative of what we consider as valid values and could confuse the reader of our intent.
When the flake8 test fails?

File: covsirphy/cleaning/jhu_complement.py (hunk 2 above)
Comment (lisphilar) [reply covsirphy/cleaning/jhu_complement.py:168]: Why the records with `confirmed < max_ignored` will be removed?
If necessary, this could be a new issue because this changes the output of `JHUData.records()` for all countries. We are currently remove only `recovered = 0` records.

Comment (Inglezos) [reply covsirphy/cleaning/jhu_complement.py:168]: I had tried that initially, because some small values (<max_ignored) led to such negative values and were problematic, but I suppose the pre_processing change shall cover for this now. I will revert this change then!

File: covsirphy/cleaning/jhu_complement.py (hunk 5 above)
Comment (lisphilar) [reply covsirphy/cleaning/jhu_complement.py:246]: Flake8/Sourcery AI say `recovery_period_is_valid = False; return recovery_period_is_valid` is unnecessary and spaces should be before/after `/`.

Please give me the details why this logic (De Morgan's laws) is confusing.

`per_up < self.upper_percentage and per_lw < self.lower_percentage` calculate True/False at the same time.

File: covsirphy/cleaning/jhu_complement.py (hunk 2 above)
Comment (lisphilar) [reply covsirphy/cleaning/jhu_complement.py:168]: OK, we will discuss this problem in a new issue because this may improve the accuracy of estimation in the 0th phase, but I think this should be handled in `JHUData` or `Scenario` class because this is not a complement.

File: covsirphy/cleaning/jhu_complement.py (hunk 5 above)
Comment (Inglezos) [reply covsirphy/cleaning/jhu_complement.py:246]: No I understand the inverse law, it is of course correct. I just wanted to give the intend that the condition is always valid, except for when we have many values over upper or under lower limit. In the inverse logic, it states that it is true if the "bad" values (either over upper or under lower) are only few and not many, which is not the first thing someone will think as comparisons. But the code is more neat I guess so I will make the suggested change.

File: covsirphy/cleaning/jhu_complement.py [hunk 7]
```diff
@@ -41,7 +45,10 @@ class JHUDataComplementHandler(Term):
         4: "fully",
     }
 
-    def __init__(self, recovery_period, interval, max_ignored, max_ending_unupdated):
+    def __init__(self, recovery_period, interval=2, max_ignored=100,
```
[the remark this answers is not in the document]
Comment (lisphilar) [reply covsirphy/cleaning/jhu_complement.py:48]: To be fixed.

[the remark this answers is not in the document]
Comment (lisphilar) [reply covsirphy/cleaning/jhu_complement.py:48]: To be fixed.

**Review by lisphilar — APPROVED** [review]
Thank you for updating!!
